harness/determined/pytorch/_metric_utils.py

```diff
@@ -0,0 +1,172 @@
+    combined_timeseries, combined_num_batches = _combine_metrics_across_processes(
+        context, metrics_timeseries, num_batches=len(per_batch_metrics)
+    )
+
+    if context.rank == 0:
+        # We can safely cast variables here because this is all happening on the chief, which
+        # is where we gather metrics.
+        combined_timeseries = cast(Dict[str, List[List[Any]]], combined_timeseries)
+        combined_num_batches = cast(List[int], combined_num_batches)
+
+        per_batch_metrics = _average_training_metrics(combined_timeseries, combined_num_batches)
+    return per_batch_metrics
+
+
+def _prepare_metrics_reducers(
+    reducer: Union[pytorch.Reducer, Dict[str, Any]], keys: Any
+) -> Dict[str, pytorch.Reducer]:
+    metrics_reducers = {}  # type: Dict[str, pytorch.Reducer]
+    if isinstance(reducer, Dict):
+        metrics_reducers = reducer
+        if keys != metrics_reducers.keys():
+            raise det.errors.InvalidExperimentException(
+                "Please provide a single evaluation reducer or "
+                "provide a reducer for every validation metric. "
+                f"Expected keys: {keys}, provided keys: {metrics_reducers.keys()}.",
+            )
+    else:
+        for key in keys:
+            metrics_reducers[key] = reducer
+
+    for key in keys:
+        if not isinstance(metrics_reducers[key], pytorch.Reducer):
+            raise det.errors.InvalidExperimentException(
+                "Please select `determined.pytorch.Reducer` for reducing validation metrics.",
+            )
+
+    return metrics_reducers
+
+
+def _convert_metrics_to_numpy(metrics: Dict[str, Any]) -> Dict[str, Any]:
+    for metric_name, metric_val in metrics.items():
+        if isinstance(metric_val, torch.Tensor):
+            metrics[metric_name] = metric_val.cpu().numpy()
+    return metrics
+
+
+def _reduce_metrics(
+    context: det._core.DistributedContext,
+    batch_metrics: List,
+    keys: Any,
+    metrics_reducers: Dict[str, pytorch.Reducer],
+) -> Dict[str, Any]:
+    metrics = {}
+    if len(batch_metrics):
+        metrics = {
+            name: pytorch._simple_reduce_metrics(
+                reducer=metrics_reducers[name],
+                metrics=np.stack([b[name] for b in batch_metrics], axis=0),
+                num_batches=None,
+            )
+            for name in keys or []
+        }
+
+    if context.size > 1:
+        # If using distributed training, combine metrics across all processes.
+        # Only the chief process will receive all the metrics.
+        num_batches = len(batch_metrics)
+        combined_metrics, batches_per_process = _combine_metrics_across_processes(
+            context, metrics, num_batches
+        )
+        if context.rank == 0:
+            # Only the chief collects all the metrics.
+            combined_metrics = _convert_metrics_to_numpy(cast(Dict[str, Any], combined_metrics))
```

Comment:
instead of a `cast`, use `assert combined_metrics is not None` which should be sufficient for typing purposes, but more useful if it ever breaks.